### PR TITLE
Fix std::abs usage in memory_optimize_pass.cc

### PR DESCRIPTION
**Pull request by GBuella**
size_t is an unsigned integer, with a conversion rank
 larger than int, therefore in the following expression
 the int value was promoted to size_t, making it a
 subtraction of unsigned values. The result of such
 a subtraction is also an unsigned value.

```
size_t space;
int space_required;
abs(space - space_required);
```

In the realm of integers, it only makes sense to use std::abs
with signed values, there is no overloaded version of std::abs
in <cstdlib> with an unsigned argument (the absolute value of an
unsigned integer is itself). The compiler has found the std::abs
template in <complex>, which resulted in the size_t value
being converted to a complex number, on which std::abs<size_t>
did operate...

See:
https://en.cppreference.com/w/cpp/language/implicit_conversion
https://en.cppreference.com/w/cpp/numeric/math/abs
https://en.cppreference.com/w/cpp/numeric/complex/abs

### CODE REVIEW

**Comment by GBuella** [comment]
By the way this fixes another build error when building using GCC 7.3

File: paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc [hunk 1]
```diff
@@ -168,7 +169,11 @@ bool FindSuitableTensorToReuse(
     if (!cluster->count(candidate)) continue;
 
     size_t space = space_table.at(candidate);
-    size_t space_diff = std::abs<size_t>(space - space_required);
+    PADDLE_ENFORCE(space <=
+                   std::numeric_limits<std::make_signed<size_t>::type>::max(),
+                   "space overload");
```
Comment (GBuella) [inline paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc]: Open for ideas for a message here, that is better than "space overload"...

**Comment by luotao1** [comment]
> By the way this fixes another build error when building using GCC 7.3

Could you paste the build error log here?

**Comment by GBuella** [comment]
> Could you paste the build error log here?

Yes, here it is (notice the reference to `/usr/include/c++/7/complex`)
```
[ 37%] Building CXX object paddle/fluid/inference/analysis/passes/CMakeFiles/memory_optim_pass.dir/memory_optimize_pass.cc.o
In file included from /home/tej/code/gbuella_paddle/build/third_party/eigen3/src/extern_eigen3/unsupported/Eigen/CXX11/../../../Eigen/Core:99:0,
                 from /home/tej/code/gbuella_paddle/build/third_party/eigen3/src/extern_eigen3/unsupported/Eigen/CXX11/Tensor:14,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/platform/device_context.h:40,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/framework/tensor.h:26,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/framework/mixed_vector.h:24,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/framework/lod_tensor.h:28,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/framework/lod_tensor_array.h:17,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/framework/var_type_traits.h:23,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/framework/variable.h:21,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/framework/scope.h:29,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/argument.h:30,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/analysis_pass.h:23,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/passes/memory_optimize_pass.h:19,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc:15:
/usr/include/c++/7/complex: In instantiation of ‘_Tp std::__complex_abs(const std::complex<_Tp>&) [with _Tp = long unsigned int]’:
/usr/include/c++/7/complex:599:56:   required from ‘_Tp std::abs(const std::complex<_Tp>&) [with _Tp = long unsigned int]’
/home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc:171:64:   required from here
/usr/include/c++/7/complex:578:35: error: call of overloaded ‘abs(long unsigned int&)’ is ambiguous
       const _Tp __s = std::max(abs(__x), abs(__y));
                                ~~~^~~~~
In file included from /usr/include/c++/7/cstdlib:77:0,
                 from /usr/include/c++/7/ext/string_conversions.h:41,
                 from /usr/include/c++/7/bits/basic_string.h:6352,
                 from /usr/include/c++/7/string:52,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/passes/memory_optimize_pass.h:16,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc:15:
/usr/include/c++/7/bits/std_abs.h:78:3: note: candidate: constexpr long double std::abs(long double)
   abs(long double __x)
   ^~~
/usr/include/c++/7/bits/std_abs.h:74:3: note: candidate: constexpr float std::abs(float)
   abs(float __x)
   ^~~
/usr/include/c++/7/bits/std_abs.h:70:3: note: candidate: constexpr double std::abs(double)
   abs(double __x)
   ^~~
/usr/include/c++/7/bits/std_abs.h:61:3: note: candidate: long long int std::abs(long long int)
   abs(long long __x) { return __builtin_llabs (__x); }
   ^~~
/usr/include/c++/7/bits/std_abs.h:56:3: note: candidate: long int std::abs(long int)
   abs(long __i) { return __builtin_labs(__i); }
   ^~~
In file included from /usr/include/c++/7/cstdlib:75:0,
                 from /usr/include/c++/7/ext/string_conversions.h:41,
                 from /usr/include/c++/7/bits/basic_string.h:6352,
                 from /usr/include/c++/7/string:52,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/passes/memory_optimize_pass.h:16,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc:15:
/usr/include/stdlib.h:837:12: note: candidate: int abs(int)
 extern int abs (int __x) __THROW __attribute__ ((__const__)) __wur;
            ^~~
In file included from /home/tej/code/gbuella_paddle/build/third_party/eigen3/src/extern_eigen3/unsupported/Eigen/CXX11/../../../Eigen/Core:99:0,
                 from /home/tej/code/gbuella_paddle/build/third_party/eigen3/src/extern_eigen3/unsupported/Eigen/CXX11/Tensor:14,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/platform/device_context.h:40,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/framework/tensor.h:26,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/framework/mixed_vector.h:24,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/framework/lod_tensor.h:28,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/framework/lod_tensor_array.h:17,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/framework/var_type_traits.h:23,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/framework/variable.h:21,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/framework/scope.h:29,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/argument.h:30,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/analysis_pass.h:23,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/passes/memory_optimize_pass.h:19,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc:15:
/usr/include/c++/7/complex:578:45: error: call of overloaded ‘abs(long unsigned int&)’ is ambiguous
       const _Tp __s = std::max(abs(__x), abs(__y));
                                          ~~~^~~~~
In file included from /usr/include/c++/7/cstdlib:77:0,
                 from /usr/include/c++/7/ext/string_conversions.h:41,
                 from /usr/include/c++/7/bits/basic_string.h:6352,
                 from /usr/include/c++/7/string:52,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/passes/memory_optimize_pass.h:16,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc:15:
/usr/include/c++/7/bits/std_abs.h:78:3: note: candidate: constexpr long double std::abs(long double)
   abs(long double __x)
   ^~~
/usr/include/c++/7/bits/std_abs.h:74:3: note: candidate: constexpr float std::abs(float)
   abs(float __x)
   ^~~
/usr/include/c++/7/bits/std_abs.h:70:3: note: candidate: constexpr double std::abs(double)
   abs(double __x)
   ^~~
/usr/include/c++/7/bits/std_abs.h:61:3: note: candidate: long long int std::abs(long long int)
   abs(long long __x) { return __builtin_llabs (__x); }
   ^~~
/usr/include/c++/7/bits/std_abs.h:56:3: note: candidate: long int std::abs(long int)
   abs(long __i) { return __builtin_labs(__i); }
   ^~~
In file included from /usr/include/c++/7/cstdlib:75:0,
                 from /usr/include/c++/7/ext/string_conversions.h:41,
                 from /usr/include/c++/7/bits/basic_string.h:6352,
                 from /usr/include/c++/7/string:52,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/passes/memory_optimize_pass.h:16,
                 from /home/tej/code/gbuella_paddle/paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc:15:
/usr/include/stdlib.h:837:12: note: candidate: int abs(int)
 extern int abs (int __x) __THROW __attribute__ ((__const__)) __wur;
            ^~~
paddle/fluid/inference/analysis/passes/CMakeFiles/memory_optim_pass.dir/build.make:62: recipe for target 'paddle/fluid/inference/analysis/passes/CMakeFiles/memory_optim_pass.dir/memory_optimize_pass.cc.o' failed

```

**Comment by tensor-tang** [comment]
Thanks @GBuella , recently we are developing based on gcc5 or gcc48, if you are trying gcc above these versions you may encounter some errors as you have seen.

If you are just trying to install or [build PaddlPaddle from source code](http://www.paddlepaddle.org/documentation/docs/en/1.2/beginners_guide/install/compile/compile_Ubuntu_en.html), you may just use docker image: paddlepaddle/paddle:latest-dev

**Review by tensor-tang — APPROVED** [review]
LGTM
Looking forward to your new PRs.

**Pull request merged**